pkg/collector/corechecks/snmp/internal/report/report_memory_usage.go

```diff
@@ -0,0 +1,300 @@
+		// memory total and memory used
+		floatMemoryFree, err := scalarMemoryFree.value.ToFloat64()
+		if err != nil {
+			log.Debugf("metric `%s`: failed to convert to float64: %s", "memory.free", err)
+			return err
+		}
+
+		floatMemoryTotal, err := scalarMemoryTotal.value.ToFloat64()
+		if err != nil {
+			log.Debugf("metric `%s`: failed to convert to float64: %s", "memory.total", err)
+			return err
+		}
+
+		memoryUsageValue, err := evaluateMemoryUsage(floatMemoryTotal-floatMemoryFree, floatMemoryTotal)
+		if err != nil {
+			return err
+		}
+
+		memoryUsageSample := MetricSample{
+			value:      valuestore.ResultValue{Value: memoryUsageValue},
+			tags:       scalarMemoryTotal.tags,
+			symbol:     checkconfig.SymbolConfig{Name: "memory.usage"},
+			options:    checkconfig.MetricsConfigOption{},
+			forcedType: "",
+		}
+
+		ms.sendMetric(memoryUsageSample)
+		return nil
+	}
+
+	// report missing dependency metrics
+	missingMetrics := []string{}
+	if !scalarMemoryUsedOk {
+		missingMetrics = append(missingMetrics, "used")
+	}
+	if !scalarMemoryFreeOk {
+		missingMetrics = append(missingMetrics, "free")
+	}
+	if !scalarMemoryTotalOk {
+		missingMetrics = append(missingMetrics, "total")
+	}
+
+	return fmt.Errorf("missing %s memory metrics, skipping scalar memory usage", strings.Join(missingMetrics, ", "))
+}
+
+func (ms *MetricSender) trySendColumnMemoryUsage(columnSamples map[string]map[string]MetricSample) error {
+	_, memoryUsageOk := columnSamples["memory.usage"]
+	if memoryUsageOk {
+		// memory usage is already sent through collected metrics
+		return nil
+	}
+
+	memoryUsedRows, memoryUsedOk := columnSamples["memory.used"]
+	memoryTotalRows, memoryTotalOk := columnSamples["memory.total"]
+	memoryFreeRows, memoryFreeOk := columnSamples["memory.free"]
+
+	if memoryUsedOk {
+		if memoryTotalOk {
+			for rowIndex, memoryUsedSample := range memoryUsedRows {
+				memoryTotalSample, memoryTotalSampleOk := memoryTotalRows[rowIndex]
+				if !memoryTotalSampleOk {
+					return fmt.Errorf("missing memory total sample at row %s, skipping memory usage evaluation", rowIndex)
+				}
+				floatMemoryTotal, err := memoryTotalSample.value.ToFloat64()
+				if err != nil {
+					return fmt.Errorf("metric `%s[%s]`: failed to convert to float64: %s", "memory.total", rowIndex, err)
+				}
+
+				floatMemoryUsed, err := memoryUsedSample.value.ToFloat64()
+				if err != nil {
+					return fmt.Errorf("metric `%s[%s]`: failed to convert to float64: %s", "memory.used", rowIndex, err)
+				}
+
+				memoryUsageValue, err := evaluateMemoryUsage(floatMemoryUsed, floatMemoryTotal)
+				if err != nil {
+					return err
+				}
+
+				memoryUsageSample := MetricSample{
+					value:      valuestore.ResultValue{Value: memoryUsageValue},
+					tags:       memoryUsedSample.tags,
+					symbol:     checkconfig.SymbolConfig{Name: "memory.usage"},
+					options:    checkconfig.MetricsConfigOption{},
+					forcedType: "",
+				}
+
+				ms.sendMetric(memoryUsageSample)
```

Comment:
It feels we can factor some logic overall:
- error handling for `*.value.ToFloat64()`
- creation of `memoryUsageSample := MetricSample{` and `ms.sendMetric(memoryUsageSample)`

that are very similar for many cases in scalar and column.

Reply:
A possible solution is to have multiple (3) methods like `sendMemoryUsageWithUsedTotal(memoryUsedSample.value, memoryTotalSample.value)` by taking directly values and then:
- handle error for `*.value.ToFloat64()`
- create MetricSample and send the metric

Same for `sendMemoryUsageWithUsedFree` and `sendMemoryUsageWithFreeTotal`

That can be used for column implementation too.

Reply:
Note that `memoryUsedSample.value` is not enough to create a sample, we need also tags coming from one of the two collected samples.

Reply:
It's probably just me, I see this as more readable than having dedicated methods for each combination. I refactored the evaluation method, which is used by all cases. Will consider your suggestion.

Reply:
The main concern is that some logic are duplicated / not very DRY:

- 12 occurrences of error handling for `*.value.ToFloat64()`
- 6 occurrences of creation of `memoryUsageSample := MetricSample{` + `ms.sendMetric(memoryUsageSample)`

There are likely differences way to factor things and https://github.com/DataDog/datadog-agent/pull/11993/files#r871122286 is just a suggestion.